File: aegis/editor/frontend/src/api/contracts.ts
import type { Domain, Project } from "@/types/domain";
function object(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value))
    throw new Error("Invalid backend response: expected an object");
  return value as Record<string, unknown>;
}
export function parseDomain(value: unknown): Domain {
  const data = object(value);
  for (const key of ["id", "name", "status", "description", "lastModified"]) {
    if (typeof data[key] !== "string")
      throw new Error(`Invalid domain response: ${key}`);
  }
  for (const key of ["roles", "obligationTypes", "codes", "rules"]) {
    if (!Array.isArray(data[key]))
      throw new Error(`Invalid domain response: ${key}`);
  }
  if (
    !["Draft", "Review", "Published", "Archived"].includes(String(data.status))
  )
    throw new Error("Unbekannter Domainstatus");
  return data as unknown as Domain;
}
export function parseProject(value: unknown): Project {
  const data = object(value);
  for (const key of ["path", "name", "loadedAt"]) {
    if (typeof data[key] !== "string")
      throw new Error(`Invalid project response: ${key}`);
  }
  if (!Array.isArray(data.domains) || !Array.isArray(data.meldFiles))
    throw new Error("Invalid project lists");
  data.domains.forEach(parseDomain);
  for (const file of data.meldFiles) {
    const item = object(file);
    if (
      typeof item.path !== "string" ||
      typeof item.name !== "string" ||
      typeof item.sizeBytes !== "number"
    )
      throw new Error("Invalid MELD file metadata");
  }
  return data as unknown as Project;
}
export function parseVerdict(value: unknown): Record<string, unknown> & {
  decision: "PERMITTED" | "FORBIDDEN" | "UNDECIDABLE";
  reasonType: string;
  revision: string;
  justificationChain: string[];
} {
  const data = object(value);
  if (
    !["PERMITTED", "FORBIDDEN", "UNDECIDABLE"].includes(
      String(data.decision),
    ) ||
    typeof data.reasonType !== "string" ||
    typeof data.revision !== "string" ||
    !Array.isArray(data.justificationChain) ||
    !data.justificationChain.every((v) => typeof v === "string")
  )
    throw new Error("Invalid Guard verdict from backend");
  return data as ReturnType<typeof parseVerdict>;
}

File: aegis/editor/contracts.py
"""Typed response projections; MELD remains the normative source of truth."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel


class ResponseModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True, extra="forbid")


class RoleResponse(ResponseModel):
    id: str
    name: str
    description: str
    rule_count: int
    obligation_type: str | None


class ObligationResponse(ResponseModel):
    id: str
    name: str
    description: str
    parent_id: str | None
    children: list[ObligationResponse]


class CodeResponse(ResponseModel):
    id: str
    name: str
    description: str
    prevalence: int


class RuleResponse(ResponseModel):
    id: str
    code: str
    agent_role: str
    modality: Literal["PERMITTED", "FORBIDDEN", "OBLIGATORY"]
    proposition: str
    specificity: int
    defeasible: bool
    source: str


class ParameterResponse(ResponseModel):
    name: str
    type: str


class DomainResponse(ResponseModel):
    id: str
    name: str
    description: str
    status: Literal["Draft", "Review", "Published", "Archived"]
    classification: Literal["publicData", "internalData", "confidentialData"]
    roles: list[RoleResponse]
    obligation_types: list[ObligationResponse]
    codes: list[CodeResponse]
    rules: list[RuleResponse]
    meld_files: list[str]
    revision: str
    code_prevalence: list[str]
    action_types: list[str]
    action_schemas: dict[str, list[ParameterResponse]]
    compile_status: Literal["ready", "failed", "uncompiled"]
    load_error: str
    last_modified: str
    conflict_count: int


class MeldFileResponse(ResponseModel):
    path: str
    name: str
    type: Literal["vocabulary", "ontology", "deontic", "inference"]
    size_bytes: int


class ProjectResponse(ResponseModel):
    path: str
    name: str
    loaded_at: str
    meld_files: list[MeldFileResponse]
    domains: list[DomainResponse]


class SourcesResponse(ResponseModel):
    revision: str
    sources: dict[str, str]


class VerdictResponse(ResponseModel):
    decision: Literal["PERMITTED", "FORBIDDEN", "UNDECIDABLE"]
    reason_type: str
    justification_chain: list[str]
    revision: str
    norms_applied: list[str]
    explanation: str
